src/main.cpp

```diff
@@ -125,15 +142,87 @@ Mat RenderFrame(void)
-		// Apply transforms here 
-		Mat rotationAroundTheSun = Mat::eye(4, 4, CV_32FC1);
-		earth.transform(rotationAroundTheSun * earthTransform);
-		moon.transform(rotationAroundTheSun * moonTransform);
-
-		// --- PUT YOUR CODE HERE ---
-		// Apply camera animation here
-	}
+		// Rotate the earth around itself
+		// 1. Derive Matrix for earth trans.
+		// 1. Derive transformed pivot.
+		// 3. Set new pivot.
+		// 2. Apply the transformation on the earth.
+        Mat eT = CTransform().rotate(Vec3f(0, 1, 0), 2.0f).get();
+        earth.setPivot(earthPivot);
+        earth.transform(eT);
+        std::cout << earthPivot;
+
+        // Rotate Moon around Earth:
+        // 1. Derive Matrix for moon trans.
+        // 1. Derive transformed pivot.
+        // 3. Set new pivot.
+        // 2. Apply the transformation on the earth.
+        Mat mT = CTransform().rotate(Vec3f(0, 1, 0),0.0529262087f).get();
+		auto moonPivot = earthPivot;
+        moon.setPivot(moonPivot);
+        moon.transform(mT);
+        std::cout << moonPivot;
+
+        // Rotate Earth around Sun:
+        // 1. Derive Matrix for moon trans.
+        // 3. Set new pivot.
+        // 2. Apply the transformation on the earth.
+        Mat sT = CTransform().rotate(Vec3f(0, 1, 0), 0.00547945205f).get();
+        earth.setPivot(sunPivot);
+        earth.transform(sT);
+        earthPivot = CTransform::point(earthPivot, sT);
+
+		// Applying camera animation here
+		// TARGET
+        // keyframe 0: target = (150'000, 0, -384);
+        // keyframe n - 1: target = (149' 978, 0, -2'603) - location of the Earth at the last frame
+        auto tarVec = tKeyframe1 - tKeyframe0;
+        auto tNorm = norm(tarVec);
+        tarVec = normalize(tarVec);
+        auto tStep = (float(frame)/nFrames);
+        auto nTarget = tKeyframe0 + tarVec * (tNorm * tStep);
+        cam3->setTarget(nTarget);
+
+		// POSITION
+        // frame 0: origin = (149'989, 3, 250)
+        // frame n / 2: origin = (149'500, -8, -1'300)
+        // frame n - 1: origin = (149'400, 3, - 2'800)
+        if (frame > 0 && frame < nFrames/2) {
+            auto dirVec = pKeyframe1 - pKeyframe0;
+            auto vNorm = norm(dirVec);
+            dirVec = normalize(dirVec);
+            auto step = (float) frame / ((float) nFrames / 2);
+            auto nextPoint = pKeyframe0 + (dirVec * (vNorm * step));
+            cam3->setPosition(nextPoint);
+        } else {
+            auto dirVec = pKeyframe2 - pKeyframe1;
+            auto vNorm = norm(dirVec);
+            dirVec = normalize(dirVec);
+            auto step = (float) frame / ((float) nFrames / 2);
+            auto nextPoint = pKeyframe1 + (dirVec * (vNorm * step));
+            cam3->setPosition(nextPoint);
+        }
+
+        // FRAME
+        // frame 0: angle = 3.5
+        // frame n / 2: angle = 60
+        // frame n - 1: angle = 30
+        if (frame > 0 && frame < nFrames/2)
+        {
+            auto step = frame/(nFrames/2);
```

Comment:
step all the time is equal to zero. You need to use
`auto step = static_cast<float>(frame)/(nFrames/2);`